src/include/brain/index_selection_util.h

```diff
@@ -0,0 +1,260 @@
+
+  /**
+   * @brief - Returns the number of indexes in the configuration
+   */
+  size_t GetIndexCount() const;
+
+  /**
+   * @brief - Returns the indexes in the configuration
+   */
+  const std::set<std::shared_ptr<IndexObject>> &GetIndexes() const;
+
+  /**
+   * @brief - Equality operator of the index configurations
+   */
+  bool operator==(const IndexConfiguration &obj) const;
+
+  /**
+   * @brief - Set difference of the two configurations
+   */
+  IndexConfiguration operator-(const IndexConfiguration &obj);
+
+  const std::string ToString() const;
+
+  void Clear();
+
+ private:
+  // The set of hypothetical indexes in the configuration
+  std::set<std::shared_ptr<IndexObject>> indexes_;
+};
+
+//===--------------------------------------------------------------------===//
+// IndexObjectPool
+//===--------------------------------------------------------------------===//
+
+// This class is a wrapper around a map from the IndexConfiguration to the
+// shared pointer of the object. This shared pointer is used else where in the
+// the algorithm to identify a configuration - memoization, enumeration,
+// equality while sorting etc.
+class IndexObjectPool {
+ public:
+  /**
+   * @brief - Constructor
+   */
+  IndexObjectPool() {}
+
+  /**
+   * @brief - Return the shared pointer of the object from the global
+   */
+  std::shared_ptr<IndexObject> GetIndexObject(IndexObject &obj);
+
+  /**
+   * @brief - Add the object to the pool of index objects
+   * if the object already exists, return the shared pointer 
+   * else create the object, add it to the pool and return the shared pointer
+   */
+  std::shared_ptr<IndexObject> PutIndexObject(IndexObject &obj);
+
+ private:
+  // The mapping from the object to the shared pointer
+  std::unordered_map<IndexObject, std::shared_ptr<IndexObject>,
+                     IndexObjectHasher>
+      map_;
+};
+
+//===--------------------------------------------------------------------===//
+// Workload
+//===--------------------------------------------------------------------===//
+
+// Represents a workload of SQL queries
+class Workload {
+ public:
+  /**
+   * @brief - Constructor
+   */
+  Workload() {}
+
+  /**
+   * @brief - Initialize a workload with the given query strings. Parse, bind and
+   * add SQLStatements.
+   */
+  Workload(std::vector<std::string> &queries, std::string database_name) {
+
+    LOG_DEBUG("Initializing workload with %ld queries", queries.size());
+
+    auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+    auto parser = parser::PostgresParser::GetInstance();
+    auto txn = txn_manager.BeginTransaction();
+
+    std::unique_ptr<binder::BindNodeVisitor> binder(
+      new binder::BindNodeVisitor(txn, database_name));
+
+    // Parse and bind every query. Store the results in the workload vector.
+    for (auto it = queries.begin(); it != queries.end(); it++) {
+      auto query = *it;
+      LOG_INFO("Query: %s", query.c_str());
```

Comment:
nit: maybe change this to debug or trace

Reply:
Done